Amcrest: Add on/off support & attributes. Bump amcrest to 1.3.0

pnbruckner opened the pull request:
## Breaking Change:
Remove `Build Date`, `Version` and `Serial Number` attributes from sensors. These values do not change during the life of the device and are not appropriate for state attributes. They might go into `device_info` if this was loaded via config entry.

## Description:
Add support for turn_on & turn_off services.

Add implementation of is_recording method, as well as brand and model attributes. Remove "static" attributes from sensors.

Bump amcrest package to 1.3.0 required for above changes and also handles errors in storage commands (see related issues.)

**Related issue (if applicable):** fixes #19982 

**Pull request in [home-assistant.io](https://github.com/home-assistant/home-assistant.io) with documentation (if applicable):** N/A

## Example entry for `configuration.yaml` (if applicable):
N/A

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [x] There is no commented out code in this PR.

If the code communicates with devices, web services, or third-party tools:
  - [x] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [x] New or updated dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard/__init__.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard/__init__.py#L23

pnbruckner committed:
Amcrest: Add on/off support & attributes to camera entity. Bump amcrest package to 1.3.0.

Add support for turn_on & turn_off services.

Add implementation of is_recording method, as well as brand, model, hardware_version, machine_name, serial_number, software_build and software_version attributes.

Bump amcrest package to 1.3.0 required for above changes and also handles errors in storage commands which resolves #19982.

andrewsayre commented on homeassistant/components/amcrest/camera.py:
```diff
@@ -92,12 +106,129 @@ def __init__(self, hass, amcrest):
+    # Additional Amcrest Camera properties
+
+    @property
+    def video_enabled(self):
```
Eliminate this property.  It's only used in `is_on` where you can simply return `self._camera.video_enabled`

pnbruckner replied:
See  comments on the corresponding setter. Also, if this property were to be removed, shouldn't is_on return self.is_streaming rather than interacting with the device?

andrewsayre replied:
It is perfectly fine for it to return a value from the device.  It will only be called when updating state.

pnbruckner replied:
`self._camera.video_enabled` doesn't retain any state from the device, but rather will communicate with the device to retrieve the information. It seems from your other feedback that this shouldn't be happening inside the `is_on` property. So I think the correct thing to do is to have `is_on` return the `self.is_streaming` attribute. Does that sound right?

andrewsayre replied:
Correct then access of `self._camera.video_enabled` should be treated like a method and called from `update`.  If `is_streaming` means the same as `is_on` then yes that's fine to return.

andrewsayre commented on homeassistant/components/amcrest/camera.py:
```diff
@@ -92,12 +106,129 @@ def __init__(self, hass, amcrest):
+        return self.is_streaming
+
+    @video_enabled.setter
+    def video_enabled(self, enable):
```
This isn't appropriate logic for property setter.  Move this to the `turn_on` and `turn_off` methods.

pnbruckner replied:
Sorry, but I don't understand why you say this (and the corresponding property) is inappropriate. I'm encapsulating the interaction with the device and the manipulation of the corresponding attribute (is_streaming.) Isn't that what properties and setters are for - i.e., encapsulation? Also, this method factors out the logic I would have to duplicate in turn_on and turn_off, which in my programming experience, would be a detrimental change.

andrewsayre replied:
Properties are expected to be fast, idempotent, safe (no exceptions, except validation) and single principle.  This implementation does may trigger a state update within HA which is far outside of these principles.  If you're concerned about duplication of code, create a private method `_set_video_state(state: bool)`.

pnbruckner replied:
Ok. Thanks again for the explanation. I'll remove the video_enabled property & setter and move the code from the setter into a private method.

andrewsayre commented on homeassistant/components/amcrest/camera.py:
```diff
@@ -92,12 +106,129 @@ def __init__(self, hass, amcrest):
+    @property
+    def supported_features(self):
+        """Flag supported features."""
+        return SUPPORT_ON_OFF
```
I think you are missing `SUPPORT_STREAM`

pnbruckner replied:
I'm not adding stream support. That was already done by someone else. But if this flag should be added, I can certainly do that.

andrewsayre replied:
Understand, but it's a flag that's missing.

pnbruckner replied:
I just saw that someone else has made that change since I created my branch. I'll merge. Thx!

andrewsayre commented on homeassistant/components/amcrest/camera.py:
```diff
@@ -92,12 +106,129 @@ def __init__(self, hass, amcrest):
+        if value is not None:
+            self._static_attrs[attr] = _extract_attr(value)
+
+    def _update_static_attrs(self):
```
We only want state attributes that users would use in automations and change during the life of the device. None of the elements added in this method belong in `device_state_attributes`.  Some would go into `device_info` if this was loaded via config entry.

pnbruckner replied:
FWIW, these attributes were already in the sensors, and each of the sensors were retrieving them on every update, resulting in a LOT of unnecessary extra network traffic. This was the first part of removing them from the sensors (which I was planning on doing in my next PR), where they certainly didn't belong. However, if you don't think they should be anywhere, I can remove them here and in sensors.py.

andrewsayre replied:
They don't belong in state attributes.  If this information needs to be exposed, it can be done through `device_info` after adding support for the device registry.

pnbruckner replied:
Makes perfect sense. Thanks for the explanation. I'll remove them from both camera.py & sensors.py.

andrewsayre commented on homeassistant/components/amcrest/camera.py:
```diff
@@ -92,12 +106,129 @@ def __init__(self, hass, amcrest):
+            if not self._static_attrs:
+                self._update_static_attrs()
+            self.is_streaming = self._camera.video_enabled
+            self._is_recording = self._camera.record_mode == 'Manual'
```
Move this to the property implementation and eliminate the backing field.

pnbruckner replied:
Now you have me confused. First, I thought you told me to remove the video_enabled property. Also I thought it was a rule that data from the device should only be retrieved in the update method. Lastly, this component mostly operates in an optimistic mode; the inherited should_poll method returns False and the update method is only called once at startup to retrieve the current values from the device. Thereafter the attributes are only updated in response to services. If I do as you suggest that would completely defeat the optimistic behavior. Which reminds me - should I add assumed_state and have it return True? (I didn't think so since the Camera class didn't, even though it's should_poll method returns False.)

andrewsayre replied:
> Also I thought it was a rule that data from the device should only be retrieved in the update method

That is not a rule I'm aware of.  This pattern is used extensively across platforms.

> Lastly, this component mostly operates in an optimistic mode; the inherited should_poll method returns False and the update method is only called once at startup to retrieve the current values from the device. Thereafter the attributes are only updated in response to services. If I do as you suggest that would completely defeat the optimistic behavior

Correct that `update` will only be called when added to HASS, but the state of the entity will be computed anytime `schedule_update_ha_state` is called, which will pull in attributes and properties of the entity.  Unless `record_mode` can only be retrieved once at startup, moving it to the property directly will not change the behavior.

pnbruckner replied:
Sorry, when I said "data from the device" I meant actually fetching data from the physical device via the network, not simply retrieving information from the device object that represents the physical device.

The `self._camera` properties `video_enabled` and `record_mode` both fetch data from the device via the network. That is how the amcrest package was written. It actually requires a bit of an overhaul for this and many other reasons, but that's outside the scope of this PR.

So, given that, if I understand correctly, the call to these property methods should stay in update and I should continue to use the `is_streaming` and `_is_recording` attributes (which I assume is what you meant by "backing field.")

andrewsayre reviewed (changes requested):
Good start!

pnbruckner commented:
@andrewsayre thanks for the feedback! See replies to your individual suggestions.

pnbruckner committed:
Update per review

Rebase to upstream/dev.

Remove video_enabled property and setter and replace with _enable_video_stream
method.

Remove static attributes from camera and sensors.

pnbruckner commented:
@andrewsayre updated per your review comments. Not sure what the check failures are about. They don't look like they have anything to do with my changes.

andrewsayre commented:
> Not sure what the check failures are about. They don't look like they have anything to do with my changes.

They don't, #22556 fixed that.  If you rebase again it should resolve.

pnbruckner commented:
@andrewsayre looks like one of the tests failed due to some kind of network glitch while installing packages. Is there a way to restart it?

pnbruckner commented:
So @andrewsayre, did I satisfy your requests and is this good to go now? Or was there something else you'd like me to do?

andrewsayre reviewed (approved):
Nice job!

MartinHjelmare commented:
Should we add breaking change label and paragraph since we removed some sensor state attributes?

andrewsayre commented:
Good catch, yes!

andrewsayre commented:
@pnbruckner Please edit the description above and enter the breaking change section.  Thanks!

pnbruckner commented:
>Please edit the description above and enter the breaking change section. Thanks!

@andrewsayre will do.

This might not be the appropriate forum to ask this question. If not, please suggest the appropriate channel. But since we're kind of already on the topic of the amcrest component and breaking changes...

I wrote a custom version of this component a while back that I and others have been using. I got quite a few requests to submit the changes to the standard component, which is what I've been doing in small chunks (of course, evolving per review comments and what I've learned since writing the custom component.) One of the next changes I'd like to submit would add the ability to turn on and off the camera's audio stream, recording, it's motion detection, as well as move the camera to a preset PTZ location and turn on and off its PTZ "tour" feature. I did that as services, with added attributes where appropriate, in my custom component.

Now here's where the question comes in. The component already has switches for turning on and off motion detection and motion recording. Personally I didn't think that was a good idea, and apparently, if I read his comment (towards the end of PR #12992, when the switches were first added) correctly, neither did @balloob. What I'd like to do is remove the switches and add the camera services and attributes instead. If someone _really wants_ a switch they can easily create a template switch that uses the appropriate service and attribute.

What do you think? And, again, if I should be asking this question elsewhere, please point me in the right direction. I'm still pretty new to this project.
